src/hotspot/share/prims/jvmtiClassFileReconstituter.cpp

```diff
@@ -511,7 +510,10 @@ void JvmtiClassFileReconstituter::write_record_attribute() {
     RecordComponent* component = components->at(i);
     write_u2(component->name_index());
     write_u2(component->descriptor_index());
-    write_u2(component->attributes_count());
+    u2 attributes_count = component->generic_signature_index() != 0 ? 1 : 0
+                        + component->annotations() != nullptr ? 1 : 0
+                        + component->type_annotations() != nullptr ? 1 : 0;
+    write_u2(attributes_count);
```

Comment:
Nit: I would suggest to add add the function to the `RecordComponent` class:
  u2 attributes_count() const {
     u2 attributes_count =  generic_signature_index() != 0 ? 1 : 0
                                      + annotations() != nullptr ? 1 : 0
                                      + type_annotations() != nullptr ? 1 : 0;
     return attributes_count;
  }

Reply:
RecordComponent contains information about Record attribute, but it knows nothing about class bytes and attributute_count doesn't make sense for it.
So I think this is JvmtiClassFileReconstituter responsibility to calculate a value required to generate class bytes for RecordComponent.

Reply:
The `attributute_count ` is a property of the `RecordComponent` even though it is not kept but calculated.
The `JvmtiClassFileReconstituter` should have a minimal knowledge about this property and how it is calculated, the same as any other consumer of the `RecordComponent`. Thre `RedefineClasses` in the future may need this number as well. Would it be also its responsibility to calculate it?

Reply:
`attributute_count` is a property of `component_info` Record's attribute in the class file (it contains length of the following `attribute_info_attributes` array).
`RecordComponent` is VM representation of Record class attribute and contains data required by VM in a form that is useful for the VM.
`RecordComponent` knows nothing `attribute_info_attributes` array, so it does not need to know its length.

 ClassFileParser parses `attribute_info_attributes` array and convert it to corresponding fields of `RecordComponent`.
`JvmtiClassFileReconstituter` performs the reverse operation.
They may have different `attributute_count` (that's what this bug about).
If `RedefineClasses` needs the value, the code may be updated (it depends on what value it needs - from original class file or from reconstituted class file)

Reply:
It does not matter what the `ClassFileParser` does.

> ` JvmtiClassFileReconstituter` performs the reverse operation.

True. It should know how to put the `attribute_count` value into the class file but it does not need to know how to calculate its value. What I do not like that in your model there is that no one single place knows how to calculate this value and all existing or potential consumers should have this knowledge.

> They may have different `attributute_count `(that's what this bug about).

The bug is that the `attributute_count` field value from the class file stored in the `RecordComponent` does not match the calculated value because the invisible attribute was not saved (was ignored). My suggestion is not to store the `attributute_count` field value from the class file as it was before. The suggestion is to place the calculating function `attributute_count()` into the `RecordComponent` class.

Reply:
> True. It should know how to put the `attribute_count` value into the class file but it does not need to know how to calculate its value. What I do not like in your model is that there is no one single place which knows how to calculate this value and the existing and potential consumers should have this knowledge.

It **must** know how to calculate the value. Because this is number of `attribute_info_attributes` attributes that follow. And only `JvmtiClassFileReconstituter` knows how many attributes it's going to write.

> The bug is that the `attributute_count` field value from the class file stored in the `RecordComponent` does not match the calculated value because the invisible attribute was not saved (was ignored).

I'd consider this as design issue with the current implementation - `JvmtiClassFileReconstituter` gets the value from external source (`RecordComponent`) and uses it without validation.
This approach is inconsistent with other `JvmtiClassFileReconstituter` code.
For other similar cases it calculates record counts in place (from size of arrays, string length, from list enumerator, etc.).

Reply:
So this makes sense. You only want the number of attributes that are applicable to the RecordComponent, and that is the three attributes that you then write out in the class file reconstituter: generic_signature_index, annotations and type_annotations.